data/virt_autotest/setup_dns_service.sh

```diff
@@ -174,48 +229,153 @@ for single_nameserver in ${nameservers_array[@]};do
+awk -v sshvar="${ssh_key_path}/id_rsa" 'done != 1 && /^#.*IdentityFile.*$/ { print "IdentityFile "sshvar"\n"; done=1 } 1' ${ssh_config_file} > ${ssh_config_temp}
+mv ${ssh_config_temp} ${ssh_config_file}
+#Insert script signature to the end of ${ssh_config_file}
+echo -e "#$0" >> ${ssh_config_file}
+
+#Try to reach all vm guests by using their dns names. Quit if any failure. 
+sleep_time=60
+sleep_count=0
 failed_count=0
-for vmguest in ${vm_guestnames_array[@]};do
-        ping -c 5 ${vmguest}
-	if [[ `echo $?` -ne 0 ]];then
-		echo -e "Connection to ${vmguest} using its dns name failed !\n" | tee -a ${setup_log_file}
-	    	failed_count=$((${failed_count} + 1))
-        else
-            	echo -e "Connection to ${vmguest} using its dns name succeeded !\n" | tee -a ${setup_log_file}
-        fi
+unset vmguest_ping_failed
+declare -a vmguest_ping_failed=""
+while [ ${sleep_count} -le ${sleep_time} ];do
+	echo -e "Try $((${sleep_count}/5 + 1)) time(s) ping\n" | tee -a ${setup_log_file}
+	vmguest_index=0
+	for vmguest in ${vm_guestnames_array[@]};do
+        	ping -c 1 ${vmguest}
+		vmguest_ping_failed[${vmguest_index}]=`echo $?`
+                vmguest_index=$((${vmguest_index} + 1))
+	done
+	failed_count=0
+	vmguest_index=0
+	for vmguest in ${vm_guestnames_array[@]};do
+		if [[ ${vmguest_ping_failed[${vmguest_index}]} -ne 0 ]];then
+        		echo -e "Connection to ${vmguest} using its dns name failed !\n" | tee -a ${setup_log_file}
+                	failed_count=$((${failed_count} + 1))
+		else
+        		echo -e "Connection to ${vmguest} using its dns name succeeded !\n" | tee -a ${setup_log_file}
+		fi
+		vmguest_index=$((${vmguest_index} + 1))
+	done
+	if [[ ${failed_count} -eq 0 ]];then
+        	echo -e "Connection to all virtual machines by using DNS names is working now.\n" | tee -a ${setup_log_file}
+        	break
+	fi
+	sleep_count=$((${sleep_count} + 5))
+        sleep 5s
 done
 
-chmod 777 ${setup_log_file}
+#Upload ssh public key to all vm guests. Quit if any failure.
+sleep_time=30
+sleep_count=0
+failed_count=0
+unset vmguest_sshcopyid_failed
+declare -a vmguest_sshcopyid_failed=""
 if [[ ${failed_count} -gt 0 ]];then
-	echo -e "DNS service does not work for at least one vm. Please pay attention !\n" | tee -a ${setup_log_file}
-	exit 1
+        echo -e "DNS service does not work for at least one vm. Please pay attention !\n" | tee -a ${setup_log_file}
+        exit 1
 else
-	echo -e "SSH Connection to all virtual machines by using DNS names is working now.\n" | tee -a ${setup_log_file}
+	echo -e "DNS service works for al vm guests !\n" | tee -a ${setup_log_file}
+	while [ ${sleep_count} -le ${sleep_time} ];do
+		echo -e "Try $((${sleep_count}/5 + 1)) time(s) ssh-copy-id\n" | tee -a ${setup_log_file}
+        	vmguest_index=0
+		for vmguest in ${vm_guestnames_array[@]};do
+			sshpass -p ${ssh_key_pass} ssh-copy-id -i ${ssh_key_path}/id_rsa.pub -f root@${vmguest}
```

Comment:
QAM has code to support visiting guest without password with similar way. Can you reuse here?

Reply:
This is just one line implementation. I can not understand why I should reuse something. But if you want to reuse, please give an example here directly.

Reply:
We have `tests/virtualization/xen/ssh_hypervisor_init.pm` and `tests/virtualization/xen/ssh_guests_init.pm` and you are welcome to use and / or extend it!